File: packages/core/src/ingest/ofx.ts
import { readFile } from "node:fs/promises";

import type { Db, Movimentacao } from "@spc-up/db";
import { movimentacao } from "@spc-up/db";

import { computeHashMovimento } from "./hash";
import {
  MOVIMENTACAO_DIRECAO,
  MOVIMENTACAO_STATUS,
  TIPO_PRESTADOR,
  type ParsedTransactionRow,
  type PrestadorContext,
} from "./types";

export { computeHashMovimento } from "./hash";

function parseOfxDate(raw: string): Date {
  const digits = raw.replace(/\D/g, "").slice(0, 8);
  if (digits.length !== 8) {
    throw new Error(`Data OFX inválida: ${raw}`);
  }
  const y = Number.parseInt(digits.slice(0, 4), 10);
  const m = Number.parseInt(digits.slice(4, 6), 10);
  const d = Number.parseInt(digits.slice(6, 8), 10);
  return new Date(Date.UTC(y, m - 1, d));
}

function directionFromAmount(amount: number): {
  direcao: ParsedTransactionRow["direcao"];
  valor: string;
} {
  if (amount >= 0) {
    return { direcao: MOVIMENTACAO_DIRECAO.ENTRADA, valor: Math.abs(amount).toFixed(2) };
  }
  return { direcao: MOVIMENTACAO_DIRECAO.SAIDA, valor: Math.abs(amount).toFixed(2) };
}

function readTag(block: string, tag: string): string | null {
  const re = new RegExp(`<${tag}>([^<\\r\\n]+)`, "i");
  const match = block.match(re);
  return match?.[1]?.trim() ?? null;
}

export function parseOfxBuffer(buffer: Buffer): ParsedTransactionRow[] {
  const content = buffer.toString("utf8");
  const rows: ParsedTransactionRow[] = [];

  const blocks = content.split(/<STMTTRN>/i).slice(1);
  for (const block of blocks) {
    const amountRaw = readTag(block, "TRNAMT");
    const dateRaw = readTag(block, "DTPOSTED");
    if (amountRaw == null || dateRaw == null) {
      continue;
    }

    const amount = Number.parseFloat(amountRaw);
    if (Number.isNaN(amount)) {
      throw new Error(`Valor OFX inválido: ${amountRaw}`);
    }

    const { direcao, valor } = directionFromAmount(amount);
    const memo = readTag(block, "MEMO");
    const payee = readTag(block, "NAME");
    const descricaoRaw = (payee || memo || "").trim();
    const fitid = readTag(block, "FITID");

    rows.push({
      dataMovimento: parseOfxDate(dateRaw),
      valor,
      descricaoRaw,
      direcao,
      nrExtratoBancario: fitid,
    });
  }

  return rows;
}

/** Parse an OFX file into normalized transaction rows. */
export async function parseOfx(pathOrBuffer: string | Buffer): Promise<ParsedTransactionRow[]> {
  const buffer =
    typeof pathOrBuffer === "string" ? await readFile(pathOrBuffer) : pathOrBuffer;
  return parseOfxBuffer(buffer);
}

/** Persist parsed rows as draft movimentacoes with deduplication hashes. */
export async function persistTransactions(
  db: Db,
  uf: string,
  exercicio: number,
  arquivoId: string,
  rows: ParsedTransactionRow[],
  prestador: PrestadorContext,
): Promise<Movimentacao[]> {
  const created: Movimentacao[] = [];
  const ufUpper = uf.toUpperCase();

  for (const row of rows) {
    const [mov] = await db
      .insert(movimentacao)
      .values({
        uf: ufUpper,
        exercicio,
        dataMovimento: row.dataMovimento.toISOString().slice(0, 10),
        valor: row.valor,
        descricaoRaw: row.descricaoRaw,
        direcao: row.direcao,
        nrExtratoBancario: row.nrExtratoBancario,
        arquivoIngestaoId: arquivoId,
        sessaoPrestacaoId: prestador.sessaoPrestacaoId,
        cnpjPrestador: prestador.cnpjPrestador,
        tipoPrestador: prestador.tipoPrestador,
        diretorioMunicipalId: prestador.diretorioMunicipalId,
        status: MOVIMENTACAO_STATUS.RASCUNHO,
        hashMovimento: computeHashMovimento(prestador.cnpjPrestador, exercicio, row),
      })
      .returning();

    if (mov) {
      created.push(mov);
    }
  }

  return created;
}

File: legacy/python/spc_up/services/ingest/ofx.py
"""OFX bank statement parsing and persistence."""

from __future__ import annotations

import hashlib
from datetime import date
from decimal import Decimal
from pathlib import Path
from typing import Any
from uuid import UUID

from ofxparse import OfxParser
from sqlalchemy.orm import Session

from spc_up.models.entities import Movimentacao, MovimentacaoDirecao, MovimentacaoStatus


def _to_date(value: Any) -> date:
    if isinstance(value, date):
        return value
    return value.date()


def _transaction_direction(amount: Decimal) -> tuple[str, Decimal]:
    if amount >= 0:
        return MovimentacaoDirecao.ENTRADA.value, abs(amount)
    return MovimentacaoDirecao.SAIDA.value, abs(amount)


def parse_ofx(path: str | Path) -> list[dict[str, Any]]:
    """Parse an OFX file into normalized transaction rows."""
    with Path(path).open("rb") as handle:
        ofx = OfxParser.parse(handle)

    rows: list[dict[str, Any]] = []
    for account in ofx.accounts:
        statement = account.statement
        if statement is None:
            continue
        for transaction in statement.transactions:
            amount = Decimal(str(transaction.amount))
            direcao, valor = _transaction_direction(amount)
            description = (transaction.payee or transaction.memo or "").strip()
            rows.append(
                {
                    "data_movimento": _to_date(transaction.date),
                    "valor": valor,
                    "descricao_raw": description,
                    "direcao": direcao,
                    "nr_extrato_bancario": transaction.id or None,
                }
            )
    return rows


def compute_hash_movimento(uf: str, exercicio: int, row: dict[str, Any]) -> str:
    """Build a deduplication hash for a parsed transaction row."""
    payload = "|".join(
        [
            uf,
            str(exercicio),
            row["data_movimento"].isoformat(),
            f"{row['valor']:.2f}",
            row["descricao_raw"],
            row["direcao"],
            row.get("nr_extrato_bancario") or "",
        ]
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def persist_transactions(
    session: Session,
    uf: str,
    exercicio: int,
    arquivo_ingestao_id: UUID,
    rows: list[dict[str, Any]],
) -> list[Movimentacao]:
    """Persist parsed rows as draft movimentacoes with deduplication hashes."""
    created: list[Movimentacao] = []
    for row in rows:
        movimentacao = Movimentacao(
            uf=uf,
            exercicio=exercicio,
            data_movimento=row["data_movimento"],
            valor=row["valor"],
            descricao_raw=row["descricao_raw"],
            direcao=row["direcao"],
            nr_extrato_bancario=row.get("nr_extrato_bancario"),
            arquivo_ingestao_id=arquivo_ingestao_id,
            status=MovimentacaoStatus.RASCUNHO.value,
            hash_movimento=compute_hash_movimento(uf, exercicio, row),
        )
        session.add(movimentacao)
        created.append(movimentacao)
    session.commit()
    return created
